lisa/sut_orchestrator/azure/notifiers.py

```diff
@@ -0,0 +1,132 @@
+    name: str = ""
+    vm_family: str = ""
+    core_count: int = 0
+
+
+@dataclass
+class EnvironmentVmsizeInfo:
+    name: str
+    status: EnvironmentStatus = EnvironmentStatus.New
+    vm_sizes: List[VmsizeBasicInfo] = field(default_factory=list)
+
+
+class AzureNotifier(notifier.Notifier):
+    """
+    This notifier is used to collect the core usage information, which can then guide
+    the quota request in the subscription.
+    """
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "azure_notifier"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return schema.Notifier
+
+    def _received_message(self, message: messages.MessageBase) -> None:
+        if isinstance(message, EnvironmentMessage):
+            self._process_environment_message(message)
+        else:
+            raise LisaException(f"unsupported message received, {type(message)}")
+
+    def _subscribed_message_type(self) -> List[Type[messages.MessageBase]]:
+        return [EnvironmentMessage]
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        env_path = constants.RUN_LOCAL_LOG_PATH / "environments"
+        env_path.mkdir(exist_ok=True, parents=True)
+        self._core_stats_file_path = env_path / "core_stats.json"
+        self._collection_file_path = env_path / "core_stats_collection.log"
+        self._environments: Dict[str, EnvironmentVmsizeInfo] = {}
+        self._max_cores: Dict[str, int] = {}
+        self._current_cores: Dict[str, int] = {}
+        self._write_header = True
+
+    def _process_environment_message(self, msg: EnvironmentMessage) -> None:
+        if (
+            msg.status != EnvironmentStatus.Deployed
+            and msg.status != EnvironmentStatus.Deleted
+        ):
+            return
+        env = self._environments.get(msg.name, None)
+        if not env:
+            env = EnvironmentVmsizeInfo(name=msg.name, status=msg.status)
+            assert msg.runbook.nodes_requirement
+            for node in msg.runbook.nodes_requirement:
+                node_runbook = node.get_extended_runbook(AzureNodeSchema, AZURE)
+                if node_runbook.resource_sku:
+                    vm_family = node_runbook.resource_sku["family"]
+                    vm_size = node_runbook.vm_size
+                    core_count = node.core_count
+                    assert isinstance(core_count, int), f"actual: {type(core_count)}"
+                    env.vm_sizes.append(
+                        VmsizeBasicInfo(
+                            name=vm_size, vm_family=vm_family, core_count=core_count
+                        )
+                    )
+            self._environments[msg.name] = env
+
+        env.status = msg.status
+        self._update_information(env)
+        self._dump_core_stats()
+
+    def _update_information(self, env: EnvironmentVmsizeInfo) -> None:
+        if self._write_header:
+            with open(self._collection_file_path, "w") as f:
```

Comment:
Write it in the general LISA log, instead of a new log file.

Reply:
I don't think so. It not easy to read if in general LISA log. This is only used in usage counter scenarios, a new log file should be fine.